# In-Class Coding Lab: Data Analysis with Pandas

In this lab, we will perform a data analysis on the **RMS Titanic** passenger list. The RMS Titanic is one of the most famous ocean liners in history. On April 15, 1912 it sank after colliding with an iceberg in the North Atlantic Ocean. To learn more, read here: https://en.wikipedia.org/wiki/RMS_Titanic 

Our goal today is to perform a data analysis on a subset of the passenger list. We're looking for insights as to which types of passengers did and didn't survive. Women? Children? 1st Class Passengers? 3rd class? Etc. 

I'm sure you've heard the expression often said during emergencies: "Women and Children first" Let's explore this data set and find out if that's true!

Before we begin you should read up on what each of the columns mean in the data dictionary. You can find this information on this page: https://www.kaggle.com/c/titanic/data 


## Loading the data set

First we load the dataset into a Pandas `DataFrame` variable. The `sample(10)` method takes a random sample of 10 passengers from the data set.

In [1]:
import pandas as pd
import numpy as np

# this turns off warning messages
import warnings
warnings.filterwarnings('ignore')

passengers = pd.read_csv('CCL-titanic.csv')
passengers.sample(15)

PassengerId  Survived  Pclass  \
97            98         1       1   
757          758         0       2   
167          168         0       3   
85            86         1       3   
283          284         1       3   
443          444         1       2   
628          629         0       3   
419          420         0       3   
536          537         0       1   
2              3         1       3   
80            81         0       3   
327          328         1       2   
113          114         0       3   
704          705         0       3   
546          547         1       2   

                                                  Name     Sex   Age  SibSp  \
97                     Greenfield, Mr. William Bertram    male  23.0      0   
757                           Bailey, Mr. Percy Andrew    male  18.0      0   
167    Skoog, Mrs. William (Anna Bernhardina Karlsson)  female  45.0      1   
85   Backstrom, Mrs. Karl Alfred (Maria Mathilda Gu...  female  33.0      3   
283                         Dorking, Mr. Edward Arthur    male  19.0      0   
443                          Reynaldo, Ms. Encarnacion  female  28.0      0   
628                          Bostandyeff, Mr. Guentcho    male  26.0      0   
419                          Van Impe, Miss. Catharina  female  10.0      0   
536                  Butt, Major. Archibald Willingham    male  45.0      0   
2                               Heikkinen, Miss. Laina  female  26.0      0   
80                                Waelens, Mr. Achille    male  22.0      0   
327                            Ball, Mrs. (Ada E Hall)  female  36.0      0   
113                            Jussila, Miss. Katriina  female  20.0      1   
704                            Hansen, Mr. Henrik Juul    male  26.0      1   
546                  Beane, Mrs. Edward (Ethel Clarke)  female  19.0      1   

     Parch            Ticket     Fare    Cabin Embarked  
97       1          PC 17759  63.3583  D10 D12        C  
757      0             29108  11.5000      NaN        S  
167      4            347088  27.9000      NaN        S  
85       0           3101278  15.8500      NaN        S  
283      0        A/5. 10482   8.0500      NaN        S  
443      0            230434  13.0000      NaN        S  
628      0            349224   7.8958      NaN        S  
419      2            345773  24.1500      NaN        S  
536      0            113050  26.5500      B38        S  
2        0  STON/O2. 3101282   7.9250      NaN        S  
80       0            345767   9.0000      NaN        S  
327      0             28551  13.0000        D        S  
113      0              4136   9.8250      NaN        S  
704      0            350025   7.8542      NaN        S  
546      0              2908  26.0000      NaN        S

## How many survived?

One of the first things we should do is figure out how many of the passengers in this data set survived. Let's start with isolating just the `'Survivied'` column into a series:

In [2]:
passengers['Survived'].sample(10)

324    0
624    0
748    0
751    1
73     0
336    0
150    0
648    0
650    0
708    1
Name: Survived, dtype: int64

There's too many to display so we just display a random sample of 10 passengers. 

- 1 means the passenger survivied
- 0 means the passenger died

What we really want is to count the number of survivors and deaths. We do this by querying the `value_counts()` of the `['Survived']` column, which returns a `Series` of counts, like this:

In [3]:
passengers['Survived'].value_counts()

0    549
1    342
Name: Survived, dtype: int64

Only 342 passengers survived, and 549 perished. Let's observe this same data as percentages of the whole. We do this by adding the `normalize=True` named argument to the `value_counts()` method.

In [4]:
passengers['Survived'].value_counts(normalize=True)

0    0.616162
1    0.383838
Name: Survived, dtype: float64

**Just 38% of passengers in this dataset survived.**

### Now you Try it!

**FIRST** Write a Pandas expression to display counts of males and female passengers using the `Sex` variable:

In [5]:
# todo write code here
passengers['Sex'].value_counts()

male      577
female    314
Name: Sex, dtype: int64

**NEXT** Write a Pandas expression to display male /female passenger counts as a percentage of the whole number of passengers in the data set.

In [6]:
# todo write code here
passengers['Sex'].value_counts(normalize=True)

male      0.647587
female    0.352413
Name: Sex, dtype: float64

If you got things working, you now know that **35% of passengers were female**.

## Who survivies? Men or Women?

We now know that 35% of the passengers were female, and 65% we male. 

**The next thing to think about is how do survivial rates affect these numbers? **

If the ratio is about the same for surviviors only, then we can conclude that your **Sex** did not play a role in your survival on the RMS Titanic. 

Let's find out.

In [7]:
survivors = passengers[passengers['Survived'] ==1]
survivors['PassengerId'].count()

342

Still **342** like we discovered originally. Now let's check the **Sex** split among survivors only:

In [8]:
survivors['Sex'].value_counts()

female    233
male      109
Name: Sex, dtype: int64

WOW! That is a huge difference! But you probably can't see it easily. Let's represent it in a `DataFrame`, so that it's easier to visualize:

In [9]:
sex_all_series = passengers['Sex'].value_counts()
sex_survivor_series = survivors['Sex'].value_counts()

sex_comparision_df = pd.DataFrame({ 'AllPassengers' : sex_all_series, 'Survivors' : sex_survivor_series })
sex_comparision_df['SexSurvivialRate'] = sex_comparision_df['Survivors'] / sex_comparision_df['AllPassengers']
sex_comparision_df

AllPassengers  Survivors  SexSurvivialRate
female            314        233          0.742038
male              577        109          0.188908

 **So, females had a 74% survival rate. Much better than the overall rate of 38%**
 
We should probably briefly explain the code above. 

- The first two lines get a series count of all passengers by Sex (male / female) and count of survivors by sex
- The third line creates a Pandas DataFrame. Recall a pandas dataframe is just a dictionary of series. We have two keys 'AllPassengers' and 'Survivors'
- The  fourth line creates a new column in the dataframe which is just the survivors / all passengers to get the rate of survival for that Sex.

## Feature Engineering: Adults and Children

Sometimes the variable we want to analyze is not readily available, but can be created from existing data. This is commonly referred to as **feature engineering**. The name comes from machine learning where we use data called *features* to predict an outcome. 

Let's create a new feature called `'AgeCat'` as follows:

- When **Age** <=18 then 'Child'
- When **Age** >18 then 'Adult'

This is easy to do in pandas. First we create the column and set all values to `np.nan` which means 'Not a number'. This is Pandas way of saying no value. Then we set the values based on the rules we set for the feature.

In [10]:
passengers['AgeCat'] = np.nan # Not a number
passengers['AgeCat'][ passengers['Age'] <=18 ] = 'Child'
passengers['AgeCat'][ passengers['Age'] > 18 ] = 'Adult'
passengers.sample(5)

PassengerId  Survived  Pclass  \
271          272         1       3   
888          889         0       3   
518          519         1       2   
102          103         0       1   
742          743         1       1   

                                                  Name     Sex   Age  SibSp  \
271                       Tornquist, Mr. William Henry    male  25.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
518  Angle, Mrs. William A (Florence "Mary" Agnes H...  female  36.0      1   
102                          White, Mr. Richard Frasar    male  21.0      0   
742              Ryerson, Miss. Susan Parker "Suzette"  female  21.0      2   

     Parch      Ticket      Fare            Cabin Embarked AgeCat  
271      0        LINE    0.0000              NaN        S  Adult  
888      2  W./C. 6607   23.4500              NaN        S    NaN  
518      0      226875   26.0000              NaN        S  Adult  
102      1       35281   77.2875              D26        S  Adult  
742      2    PC 17608  262.3750  B57 B59 B63 B66        C  Adult

Let's get the count and distrubutions of Adults and Children on the passenger list.

In [11]:
passengers['AgeCat'].value_counts()

Adult    575
Child    139
Name: AgeCat, dtype: int64

And here's the percentage as a whole:

In [12]:
passengers['AgeCat'].value_counts(normalize=True)

Adult    0.805322
Child    0.194678
Name: AgeCat, dtype: float64

So close to **80%** of the passengers were adults. Once again let's look at the ratio of `AgeCat` for survivors only. If your age has no bearing of survivial, then the rates should be the same. 

Here are the counts of Adult / Children among the survivors only:

In [13]:
survivors = passengers[passengers['Survived'] ==1]
survivors['AgeCat'].value_counts()

Adult    220
Child     70
Name: AgeCat, dtype: int64

### Now You Try it!

Calculate the `AgeCat` survival rate, similar to how we did for the `SexSurvivalRate`. 

In [15]:
agecat_all_series = passengers['AgeCat'].value_counts()
agecat_survivor_series = survivors['AgeCat'].value_counts()

# todo make a data frame, add AgeCatSurvivialRate column, display dataframe 

age_comparision_df = pd.DataFrame({ 'AllPassengers' : agecat_all_series, 'Survivors' : agecat_survivor_series })
age_comparision_df['AgeSurvivialRate'] = age_comparision_df['Survivors'] / age_comparision_df['AllPassengers']
age_comparision_df

AllPassengers  Survivors  AgeSurvivialRate
Adult            575        220          0.382609
Child            139         70          0.503597

**So, children had a 50% survival rate, better than the overall rate of 38%**

## So, women and children first?

It looks like the RMS really did have the motto: "Women and Children First."

Here are our insights. We know:

- If you were a passenger, you had a 38% chance of survival.
- If you were a female passenger, you had a 74% chance of survival.
- If you were a child passenger, you had a 50% chance of survival. 


### Now you try it for Passenger Class

Repeat this process for `Pclass` The passenger class variable. Display the survival rates for each passenger class. What does the information tell you about passenger class and survival rates?

I'll give you a hint... "Money Talks"


In [60]:
# todo: repeat the analysis in the previous cell for Pclass 
class_all_series = passengers['Pclass'].value_counts()
class_survivor_series = survivors['Pclass'].value_counts()

class_comparision_df = pd.DataFrame({ 'AllPassengers' : class_all_series, 'Survivors' : class_survivor_series })
class_comparision_df['ClassSurvivialRate'] = class_comparision_df['Survivors'] / class_comparision_df['AllPassengers']
class_comparision_df


AllPassengers  Survivors  ClassSurvivialRate
1            216        136            0.629630
2            184         87            0.472826
3            491        119            0.242363